src/data-model-providers/codedriven/CodeDrivenDataModelProvider.h

```diff
@@ -0,0 +1,109 @@
+
+    DataModel::ActionReturnStatus ReadAttribute(const DataModel::ReadAttributeRequest & request,
+                                                AttributeValueEncoder & encoder) override;
+    DataModel::ActionReturnStatus WriteAttribute(const DataModel::WriteAttributeRequest & request,
+                                                 AttributeValueDecoder & decoder) override;
+
+    void ListAttributeWriteNotification(const ConcreteAttributePath & aPath, DataModel::ListWriteOperation opType) override;
+    std::optional<DataModel::ActionReturnStatus> InvokeCommand(const DataModel::InvokeRequest & request,
+                                                               TLV::TLVReader & input_arguments, CommandHandler * handler) override;
+
+    void SetPersistentStorageDelegate(PersistentStorageDelegate * delegate) { mPersistentStorageDelegate = delegate; }
+    PersistentStorageDelegate * GetPersistentStorageDelegate() { return mPersistentStorageDelegate; }
+
+    /* Attribute tree iteration */
+    CHIP_ERROR Endpoints(ReadOnlyBufferBuilder<DataModel::EndpointEntry> & out) override;
+    CHIP_ERROR SemanticTags(EndpointId endpointId,
+                            ReadOnlyBufferBuilder<Clusters::Descriptor::Structs::SemanticTagStruct::Type> & out) override;
+    CHIP_ERROR DeviceTypes(EndpointId endpointId, ReadOnlyBufferBuilder<DataModel::DeviceTypeEntry> & out) override;
+    CHIP_ERROR ClientClusters(EndpointId endpointId, ReadOnlyBufferBuilder<ClusterId> & out) override;
+    CHIP_ERROR ServerClusters(EndpointId endpointId, ReadOnlyBufferBuilder<DataModel::ServerClusterEntry> & out) override;
+    CHIP_ERROR GeneratedCommands(const ConcreteClusterPath & path, ReadOnlyBufferBuilder<CommandId> & out) override;
+    CHIP_ERROR AcceptedCommands(const ConcreteClusterPath & path,
+                                ReadOnlyBufferBuilder<DataModel::AcceptedCommandEntry> & out) override;
+    CHIP_ERROR Attributes(const ConcreteClusterPath & path, ReadOnlyBufferBuilder<DataModel::AttributeEntry> & builder) override;
+
+    void Temporary_ReportAttributeChanged(const AttributePathParams & path) override;
+
+    /* Lifecycle Management:
+     * The CodeDrivenDataModelProvider stores pointers to EndpointProviderInterface, but does NOT take ownership.
+     * It is crucial that the lifetime of any EndpointProviderInterface instance outlives the
+     * CodeDrivenDataModelProvider it is registered with. Similarly, EndpointProviderRegistration objects
+     * passed to AddEndpoint must outlive their registration.
+     */
+    CHIP_ERROR AddEndpoint(EndpointProviderRegistration & registration);
+    CHIP_ERROR RemoveEndpoint(EndpointId endpointId);
+
+private:
+    EndpointProviderInterface * GetEndpointProvider(EndpointId endpointId);
+    EndpointProviderRegistry mEndpointProviderRegistry;
+    std::optional<ServerClusterContext> mServerClusterContext;
+    PersistentStorageDelegate * mPersistentStorageDelegate = nullptr;
```

Comment:
why is this separate from `mServerClusterContext->storage` ?

This needs explanation and feels off... can this have a different value than the context?